tensorboard/webapp/widgets/line_chart_v2/lib/renderer/renderer_types.ts

```diff
@@ -0,0 +1,39 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Paths, Rect} from '../types';
+
+export interface IRenderer {
+  /**
+   * Certain renderer requires DOM dimensions for correct density and operations. The
+   * method is invoked when container is resized.
+   *
+   * @param domRect Container dimensions
+   */
+  onResize(domRect: Rect): void;
+
+  drawLine(cacheId: string, paths: Paths, spec: LineSpec): void;
+
+  flush(): void;
+
+  renderGroup(groupName: string, renderBlock: () => void): void;
```

Comment:
I can guess from the name and type signature what “kind” of thing this
is (a scoped bracketing operation), but what are the semantics? Is this…
composited onto an intermediate buffer so that opacity transforms can be
applied atomically? rendered transactionally with opportunities for
internal performance improvements? just a helper for `flush`? Why would
I call this method, what must I ensure as a caller, and what does the
method guarantee?

What, also, are the semantics of the `groupName`? From the abstract
class, it looks like a key. Do the keys last just for the duration of
the render group, or are they owned externally to the callback? Is it
valid to reuse the key, like

```typescript
r.renderGroup("foo", () => {
  r.renderGroup("foo", () => {
  });
});
```

? If not, if the key is meant to always be unique, would it make more
sense for `renderGroup` to gensym a group name and pass it to the
`renderBlock` callback? Surely the answer depends on how the group name
is used, so… how is it used?

Reply:
Ah, you cannot nest a renderGroup so added some check and test for that. Thanks for catching that.